### ISSUE TITLE
[Helm/Release] Panic after changing values

### ISSUE THREAD

**Issue by blampe** [pulumi/pulumi-kubernetes#1780]
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

I successfully deployed a Release resource introduced in #1677. I then modified `values` in such a way that I expected it to remove 2 `DaemonSet` definitions. However when previewing it with those new values, it consistently fails with this panic:

```
$ pulumi preview
...
   panic: fatal: An assertion has failed
    goroutine 212 [running]:
    github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.14.0/go/common/util/contract/failfast.go:23
    github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Assert(...)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.14.0/go/common/util/contract/assert.go:26
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchValueToDiff(0xc002de4f78, 0xc003e3b3e0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:2835 +0xb6b
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchMapToDiff(0xc002de4f78, 0xc003e463f0, 0x1, 0x1, 0xc003dea720, 0xc003e16840, 0xc003e17d10, 0xc003e2f260, 0xc003e46300, 0x10, ...)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:2935 +0x31d
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchValueToDiff(0xc002de4f78, 0xc003e463f0, 0x1, 0x1, 0x2da92c0, 0xc003dea720, 0x2da92c0, 0xc003e16840, 0x2da92c0, 0xc003e17d10, ...)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:2856 +0x17a
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchMapToDiff(0xc002de4f78, 0x0, 0x0, 0x0, 0xc003dea6f0, 0xc003e167e0, 0xc003e17cb0, 0xc003e2f200, 0x0, 0xc002de4f98, ...)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:2935 +0x31d
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.convertPatchToDiff(0xc003dea6f0, 0xc003e167e0, 0xc003e17cb0, 0xc003e2f200, 0xc003e3b2e0, 0x2, 0x2, 0x2, 0x200, 0x0)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:2762 +0x119
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*helmReleaseProvider).Diff(0xc00075dda0, 0x346f448, 0xc003a9c390, 0xc002dc8000, 0x7, 0xc001fee25d, 0x2)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/helm_release.go:683 +0xbea
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*kubeProvider).Diff(0xc000702600, 0x346f448, 0xc003a9c390, 0xc002dc8000, 0xc000702600, 0x2dc6e01, 0xc003a9f200)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:1431 +0x6dc
    github.com/pulumi/pulumi/sdk/v3/proto/go._ResourceProvider_Diff_Handler.func1(0x346f448, 0xc003a9c390, 0x2fbbbe0, 0xc002dc8000, 0x2fd3b60, 0x464f5c8, 0x346f448, 0xc003a9c390)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.14.0/proto/go/provider.pb.go:2584 +0x89
    github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x346f448, 0xc002db23c0, 0x2fbbbe0, 0xc002dc8000, 0xc002da4d40, 0xc0006c9548, 0x0, 0x0, 0x3424e20, 0xc000342f20)
    	/home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x30a
    github.com/pulumi/pulumi/sdk/v3/proto/go._ResourceProvider_Diff_Handler(0x3076a20, 0xc000702600, 0x346f448, 0xc002db23c0, 0xc002db02a0, 0xc000a6b960, 0x346f448, 0xc002db23c0, 0xc002dbe000, 0x8780)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.14.0/proto/go/provider.pb.go:2586 +0x150
    google.golang.org/grpc.(*Server).processUnaryRPC(0xc00043dc00, 0x348c298, 0xc000682480, 0xc003a50d80, 0xc000a6fce0, 0x45ebf68, 0x0, 0x0, 0x0)
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:1286 +0x52b
    google.golang.org/grpc.(*Server).handleStream(0xc00043dc00, 0x348c298, 0xc000682480, 0xc003a50d80, 0x0)
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:1609 +0xd0c
    google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc0003e4050, 0xc00043dc00, 0x348c298, 0xc000682480, 0xc003a50d80)
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:934 +0xab
    created by google.golang.org/grpc.(*Server).serveStreams.func1
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:932 +0x1fd
```

Restoring the values to what I had previously allows it to succeed.

**Comment by viveklak** [comment]
@blampe thanks for opening the issue. Would you be able to share some of the code before and after the change that triggered the panic?

**Comment by blampe** [comment]
@viveklak I lost some state so I'm not able to easily repro right now but here's as much context as I can share:

The cluster was an EKS instance, mapped to a `k8s.Provider`:

```
    const cluster = new eks.Cluster("cluster", {
       vpcId: vpc.id,
       enabledClusterLogTypes: ["api", "audit", "scheduler"],
       publicSubnetIds: vpc.publicSubnetIds,
       instanceType: "t2.medium",
       desiredCapacity: 2,
       minSize: 1,
       maxSize: 5,
       createOidcProvider: true,
       version: "1.21",
       endpointPrivateAccess: true,
       endpointPublicAccess: true,
     });

    provider = new k8s.Provider("provider", {
      kubeconfig: cluster.kubeconfig,
      helmReleaseSettings: {
      suppressBetaWarning: true,
      driver: "secret",
   },
```

The Release was for the Jupyterhub helm chart. This is what the initial values looked like, minus some configuration which shouldn't impact anything:
```
new k8s.helm.v3.Release(
  "jupyter-hub",
  {
    version: "1.1.2",
    chart: "jupyterhub",
    repositoryOpts: {
      repo: "https://jupyterhub.github.io/helm-chart/",
    },
    values: {
      debug: {
        enabled: true,
      },
      prePuller: {
        hook: {
          enabled: false,
        },
      },
      singleuser: {
        image: {
          name: "busybox", // Dummy data
        },
        defaultUrl: "/lab",
        cmd: null,
        startTimeout: 3600,
      },
      proxy: {
        https: {
          enabled: true,
          type: "offload",
        },
      },
    },
  },
  { provider: provider },
);
```
I was able to successfully deploy something with roughly these values. Panic resulted after changing `prePuller` to 
```
prePuller: {
        hook: {
          enabled: false,
        },
        continuous: {
          enabled: false,
        },
      },
```

which should have had the effect of removing a daemonset.

**Comment by viveklak** [comment]
@blampe thanks for sharing as much context as you could. I will try to replicate and update.

**Comment by viveklak** [comment]
@blampe I tried reproducing this on my end but I haven't been successful. Tracing through the stacktrace, essentially the panic is due to an[ assertion failure](https://github.com/pulumi/pulumi-kubernetes/blob/ff43b7c3b49670404b1642e0f7ca1213efe4e5c2/provider/pkg/provider/provider.go#L2835) where we generate a patch to execute a change (based on user input) where we expect a diff for a field but discover that the corresponding field is missing in both the old and new state. We could potentially relax this to be a no-op but I can't really identify how we can get into this state. 

If you have a state dump of the stack as of the failure, that would be really helpful (`pulumi stack history` followed by a `pulumi stack export --version $version`).

**Comment by viveklak** [comment]
For now I will drop the emergent label and priority since I can't easily reproduce this.

**Comment by ruckc** [comment]
I keep running into this when I am creating a new k8s resource and forget to specify the resource `provider` argument manually in the `CustomResourceOptions` argument.

**Comment by viveklak** [comment]
@ruckc could you provide the stacktrace and code you are using? Thanks in advance!

**Comment by ruckc** [comment]
The stacktrace:

    panic: fatal: An assertion has failed
    goroutine 13986 [running]:
    github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.19.0/go/common/util/contract/failfast.go:23
    github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Assert(...)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.19.0/go/common/util/contract/assert.go:26
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchValueToDiff(0xc00404af78, 0xc0051e1680, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:2850 +0xb6b
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchMapToDiff(0xc00404af78, 0xc0051f2fd0, 0x1, 0x1, 0xc0051b9e30, 0xc0051dd4d0, 0xc0051ddb90, 0xc0051f6300, 0xc0051f2f00, 0x10, ...)
        /home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:2950 +0x31d
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchValueToDiff(0xc00404af78, 0xc0051f2fd0, 0x1, 0x1, 0x21b1500, 0xc0051b9e30, 0x21b1500, 0xc0051dd4d0, 0x21b1500, 0xc0051ddb90, ...)
        /home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:2871 +0x17a
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchMapToDiff(0xc00404af78, 0x0, 0x0, 0x0, 0xc0051b9e00, 0xc0051dd470, 0xc0051ddb30, 0xc0051f62a0, 0x0, 0xc00404af98, ...)
        /home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:2950 +0x31d
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.convertPatchToDiff(0xc0051b9e00, 0xc0051dd470, 0xc0051ddb30, 0xc0051f62a0, 0xc0051e1660, 0x2, 0x2, 0x2, 0x480, 0x0)
        /home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:2777 +0x119
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*helmReleaseProvider).Diff(0xc0006968a0, 0x2897b48, 0xc005088a20, 0xc00503bdc0, 0x7, 0xc00504442c, 0x2)
        /home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/helm_release.go:687 +0xbea
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*kubeProvider).Diff(0xc000582f00, 0x2897b48, 0xc005088a20, 0xc00503bdc0, 0xc000582f00, 0x21cf001, 0xc00508d2c0)
        /home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:1447 +0x6dc
    github.com/pulumi/pulumi/sdk/v3/proto/go._ResourceProvider_Diff_Handler.func1(0x2897b48, 0xc005088a20, 0x23c3b20, 0xc00503bdc0, 0x23db960, 0x3ab62a8, 0x2897b48, 0xc005088a20)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.19.0/proto/go/provider.pb.go:2584 +0x89
    github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x2897b48, 0xc005064270, 0x23c3b20, 0xc00503bdc0, 0xc00502d8c0, 0xc00506a0c0, 0x0, 0x0, 0x284d6a0, 0xc00033ee60)
        /home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x30a
    github.com/pulumi/pulumi/sdk/v3/proto/go._ResourceProvider_Diff_Handler(0x247e9a0, 0xc000582f00, 0x2897b48, 0xc005064270, 0xc005069080, 0xc000a34660, 0x2897b48, 0xc005064270, 0xc00507c000, 0x3ec5)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.19.0/proto/go/provider.pb.go:2586 +0x150
    google.golang.org/grpc.(*Server).processUnaryRPC(0xc000263c00, 0x28b48f8, 0xc000602300, 0xc004a5cd80, 0xc000a48480, 0x3a52728, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:1286 +0x52b
    google.golang.org/grpc.(*Server).handleStream(0xc000263c00, 0x28b48f8, 0xc000602300, 0xc004a5cd80, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:1609 +0xd0c
    google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc0005dc040, 0xc000263c00, 0x28b48f8, 0xc000602300, 0xc004a5cd80)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:934 +0xab
    created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:932 +0x1fd

**Comment by viveklak** [comment]
Good to know. I will look through, thanks @ruckc

**Comment by ruckc** [comment]
@viveklak, I updated my previous comment with some debug detail.

**Comment by viveklak** [comment]
#1848 seems related as well.

**Comment by viveklak** [comment]
#1850 has been merged with a fix that should address this. It will be available in 3.13.1 release next week.
